### ISSUE TITLE
ENH: speeding up model training by moving it to GPU?

### ISSUE THREAD

**Issue by vss888** [statsmodels/statsmodels#4752]
I am searching for a way to speed up model training since we need to train many of them (i.e. dozens of millions). To profile things up, I took an example model, which takes 190 seconds to train. Out of those 75% are spent in kalman_filter.py:733(_filter) and the next 17% are spent in numpy.core.multiarray.dot.

Looking at statsmodels/tsa/kalmanf/kalman_loglike.pyx, I can see that the code is already heavily optimized for CPU by using Fortran BLAS libraries. So one idea is to move at least the two functions to run on GPUs instead of CPUs.

Please, let me know if you have a feeling of much it could help and which Python CUDA library (CuPy, Numba, MinPy, PyTorch, TensorFlow, CULA, scikit-cuda, PyCUDA, PyOpenCL, pyculib, theano, etc) would be a good choice to proceed?

Thank you for your help!

**Comment by josef-pkt** [comment]
(I never worked on a GPU.)

I would say that moving ARIMA and the statespace models to GPU is close to impossible or would require a large amount of work even if LAPACK is available on GPU.

First ARIMA is mostly in minimum maintenance mode, and all new development goes into the new statespace models like SARIMAX.

The kalman filter code underlying the statespace model is tightly integrated with scipy, i.e. it uses the BLAS/LAPACK cython wrappers that scipy provides and avoids that we have to work with a specific BLAS/LAPACK library. The scipy developers did the work of providing a common interface to whatever underlying linear algebra library is used, currently mainly MKL and OpenBLAS.

I don't know if the required function from BLAS/LAPACK are available on GPU. If they are, then I think we would need a wrapper similar to what scipy provides to hook our cython kalman filter code into it. 
The other part would be that numpy functions like dot also would have to be replaced. 

Overall, I guess this would require a lot of backend work similar to DASK where none of our developers has a comparative advantage.

---
Aside to the application:
How many observations do you have per time series? 190 seconds sounds pretty long to me.
Are the time series and model specification similar across time series so you can reuse start_params and avoid specification search?

**Comment by vss888** [comment]
MinPy has interface identical to the one of NumPy and so the 2nd function should be doable to replace. Checking what BLAS functions are used in kalman_loglike.pyx: it is basically only dot product and addition of matrices, vector dot product, and matrix vector product and addition, and that is it. All of the functions are basic tensor operations, for which multiple heavily optimized GPU implementations exist, and so the project does not look impossible.

Number of points per series could vary, but in the test case above it was about 13000. We do use SARIMAX and do not use ARIMA. I did try to reuse start_params from very similar models to speed things up and improve model convergence (sometimes fits diverge), but for some reason it degrades model performance (precision of predictions gets worse), so at the moment we start training from scratch every time.

**Comment by vss888** [comment]
Could you please let me know which command can force recompilation of `statsmodels/tsa/kalmanf/kalman_loglike.cpython-*-x86_64-linux-gnu.so` with a custom local version of `kalman_loglike.pyx` ? Thank you for your help!
